include/boost/gil/image_processing/numeric.hpp

```diff
@@ -68,139 +72,218 @@ inline void compute_tensor_entries(
-        pixel.at(std::integral_constant<int, 0>{}) = 1.0f;
+    kernel_2d<T, Allocator> result(side_length, side_length / 2, side_length / 2);
+    for (auto& cell: result) {
+        cell = 1.0f;
     }
+
+    return result;
 }
 
 /// \brief Generate Gaussian kernel
 /// \ingroup ImageProcessingMath
 ///
 /// Fills supplied view with values taken from Gaussian distribution. See
 /// https://en.wikipedia.org/wiki/Gaussian_blur
-inline void generate_gaussian_kernel(boost::gil::gray32f_view_t dst, double sigma)
+template <typename T = float, typename Allocator = std::allocator<T>>
+inline kernel_2d<T, Allocator> generate_gaussian_kernel(std::size_t side_length, double sigma)
 {
-    if (dst.width() != dst.height() || dst.width() % 2 != 1)
+    if (side_length % 2 != 1)
         throw std::invalid_argument("kernel dimensions should be odd and equal");
 
+
     const double denominator = 2 * boost::gil::pi * sigma * sigma;
-    const auto middle = boost::gil::point_t(dst.width() / 2, dst.height() / 2);
-    for (boost::gil::gray32f_view_t::coord_t y = 0; y < dst.height(); ++y)
+    auto middle = side_length / 2;
+    std::vector<T, Allocator> values(side_length * side_length);
+    for (std::size_t y = 0; y < side_length; ++y)
     {
-        for (boost::gil::gray32f_view_t::coord_t x = 0; x < dst.width(); ++x)
+        for (std::size_t x = 0; x < side_length; ++x)
         {
-            const auto delta_x = std::abs(middle.x - x);
-            const auto delta_y = std::abs(middle.y - y);
+            const auto delta_x = middle > x ? middle - x : x - middle;
+            const auto delta_y = middle > y ? middle - y : y - middle;
             const double power = (delta_x * delta_x +  delta_y * delta_y) / (2 * sigma * sigma);
             const double nominator = std::exp(-power);
-            const float value = nominator / denominator;
-            dst(x, y).at(std::integral_constant<int, 0>{}) = value;
+            const float value = static_cast<float>(nominator / denominator);
+            values[y * side_length + x] = value;
+        }
+    }
+
+    return kernel_2d<T, Allocator>(values.begin(), values.size(), middle, middle);
+}
+
+/// \brief Generates Sobel operator in horizontal direction
+/// \ingroup ImageProcessingMath
+///
+/// Generates a kernel which will represent Sobel operator in
+/// horizontal direction of specified degree (no need to convolve multiple times
+/// to obtain the desired degree).
+/// https://www.researchgate.net/publication/239398674_An_Isotropic_3_3_Image_Gradient_Operator
+template <typename T = float, typename Allocator = std::allocator<T>>
+inline kernel_2d<T, Allocator> generate_dx_sobel(unsigned int degree = 1)
+{
+    switch (degree)
+    {
+        case 0:
+        {
+            return get_identity_kernel<T, Allocator>();
+        }
+        case 1:
+        {
+            kernel_2d<T, Allocator> result(3, 1, 1);
+            std::copy(dx_sobel.begin(), dx_sobel.end(), result.begin());
```

Comment:
Since we have compile-time known `std::array<float, 9> dx_sobel`, this place seems like a good opportunity to add `static_assert` verifying `dx_sobel.size()` and `BOOST_ASSERT` verifying `dx_sobel.size() == result.size()`.

Such assertions would also serve as code comment for a reader who doesn't have to jump to `dx_sobel` definition.